File: lectrick/tiles/conditional_horizontal_wire.py
from .base import Tile
from ..fire import FireException


class ConditionalHorizontalWire(Tile):
    TILE_TYPE = "CONDITIONAL HORIZONTAL WIRE"
    SAMPLE_CHARS = ["T"]

    def __init__(self, x, y, tiles):
        super().__init__(x, y, tiles)
        self.pending_vertical_input = None
        self.current_vertical_input = None

    def accept_energy(self, amount, source_x, source_y):
        if source_y == self.y:  # Horizontal input
            if self.pending_input is not None:
                raise FireException(
                    f"FIRE: Tile {self.TILE_TYPE} at {self.x}, {self.y} already has pending input ({self.pending_input}). Source: ({source_x}, {source_y})")
            self.pending_input = (amount, source_x, source_y)
        elif source_x == self.x:  # Vertical input
            if self.pending_vertical_input is not None:
                raise FireException(
                    f"FIRE: Tile {self.TILE_TYPE} at {self.x}, {self.y} already has pending vertical input ({self.pending_vertical_input}). Source: ({source_x}, {source_y})")
            self.pending_vertical_input = (amount, source_x, source_y)

    def advance_turn(self, *args, **kwargs):
        self.current_input = self.pending_input
        self.pending_input = None
        self.current_vertical_input = self.pending_vertical_input
        self.pending_vertical_input = None

    def start_turn(self, *args, **kwargs):
        if self.current_input and self.current_vertical_input:
            vertical_amount = self.current_vertical_input[0]
            if vertical_amount <= 0:
                return
            amount, source_x, source_y = self.current_input
            if source_x < self.x:
                dx = -1
            elif source_x > self.x:
                dx = 1
            else:
                return
            try:
                tile_index = self.tiles.index((self.x - dx, self.y))
            except ValueError:
                return
            self.tiles[tile_index].accept_energy(amount, self.x, self.y)

File: web/src/tiles/conditionalHorizontalWire.ts
import {Tile} from './base';
import {Input} from './oneWayBase';

export class ConditionalHorizontalWireTile extends Tile {
    TILE_TYPE = 'CONDITIONAL HORIZONTAL WIRE';
    SAMPLE_CHARS = [
        "T"
    ];
    pendingInput: Input | null;
    currentInput: Input | null;
    pendingVerticalInput: number | null;
    currentVerticalInput: number | null;

    constructor(ctx: ExecutionContext, x: number, y: number) {
        super(ctx, x, y);
        this.pendingVerticalInput = null;
        this.currentVerticalInput = null;
    }

    acceptEnergy(amount: number, x: number, y: number) {
        if (y === this.y) {
            // Horizontal input
            if (this.pendingInput !== null) {
                // Already has a pending input
                this.ctx.startFire();
                return;
            }
            this.pendingInput = [amount, x, y];
        } else if (x === this.x) {
            // Vertical input
            if (this.pendingVerticalInput !== null) {
                // Already has a pending vertical input
                this.ctx.startFire();
                return;
            }
            this.pendingVerticalInput = amount;
        }
    }

    advanceTurn() {
        this.currentInput = this.pendingInput;
        this.pendingInput = null;
        this.currentVerticalInput = this.pendingVerticalInput;
        this.pendingVerticalInput = null;
    }

    startTurn() {
        if (this.currentInput !== null && this.currentVerticalInput !== null) {
            if (this.currentVerticalInput <= 0) {
                return;
            }
            const [amount, x, y] = this.currentInput;
            let dx;
            if (x < this.x) {
                dx = -1; // Left
            } else if (x > this.x) {
                dx = 1; // Right
            } else {
                return; // No horizontal movement
            }
            const targetTile = this.ctx.getTile(this.x + dx, this.y);
            if (targetTile) {
                targetTile.acceptEnergy(amount, this.x, this.y);
            }
        }
    }
}